### PR TITLE
Option to do single precision mesh communication.

### PR DESCRIPTION
**Pull request by atmyers**
This PR adds a number of wrapper functions around the amrex parallel communication routines for mesh data - FillBoundary, ParallelCopy, ParallelAdd, SumBoundary, and OverrideSync, and replaces all uses of these functions with the wrapped versions. It also adds a new runtime option, `warpx.do_single_precision_comms` (default `0`). If this is `0`, the wrapped versions will behave as they currently do on `development`. If this is `1`, the communication will be done using a temporary, single precision version of Multifab instead. Thus, at the cost of some local copies, half as much data will need to be communicated with MPI. This is useful for simulations that use the PSATD solver and thus need a lot of ghost cells.

# To Do

- [x] do not copy data if we already built in single precision
- [x] add docs
- [x] investigate tests that crash from #2319
- [x] add tests
- [x] check if tests fail by a large margin if this would be default-on: #2319

### CODE REVIEW

File: Source/Parallelization/WarpXCommUtil.H [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        if (bx.ok())
+        {
+            auto const srcFab = src.array(mfi);
+            auto       dstFab = dst.array(mfi);
+            AMREX_HOST_DEVICE_PARALLEL_FOR_4D_FUSIBLE ( bx, numcomp, i, j, k, n,
+            {
+                dstFab(i,j,k,dstcomp+n) = srcFab(i,j,k,srccomp+n);
+            });
+        }
+    }
```
Comment (WeiqunZhang) [inline Source/Parallelization/WarpXCommUtil.H]:
```suggestion
    auto const& srcma = src.const_arrays();
    auto const& dstma = dst.arrays();
    ParallelFor(dst, nghost, numcomp,
    [=] AMREX_GPU_DEVICE (int box_no, int i, int j, int k, int n) noexcept
    {
        dstma[box_no](i,j,k,dstcomp+n) = srcma[box_no](i,j,k,srccomp+n);
    });
    Gpu::synchronize();
```

**Comment by ax3l** [comment]
In #2319, I see some tests change precision with single precision comm (as expected), but I also see quite a lot of tests crashing.

We should investigate the crashes before merging the PR here.

**Comment by ax3l** [comment]
I rebased the experiment in #2319 one last time so we see if no new crashes slurped in

File: Source/Parallelization/WarpXCommUtil.H [hunk 2]
```diff
@@ -0,0 +1,88 @@
+#ifndef WARPX_COMMUTIL_H_
```
Comment (ax3l) [inline Source/Parallelization/WarpXCommUtil.H:7]: Copyright header missing :)

**Review by ax3l — COMMENTED** [review]
Just a minimal comment on a header in a new file.

Otherwise let's wait for CI in this and the linked PR and check it one last time.

**Comment by ax3l** [comment]
You probably saw that we have 3 tests that crash:
https://github.com/ECP-WarpX/WarpX/pull/2319#issuecomment-930350038

They all seem to be related to space charge, which we don't use in the MS.

**Comment by atmyers** [comment]
> You probably saw that we have 3 tests that crash: [#2319 (comment)](https://github.com/ECP-WarpX/WarpX/pull/2319#issuecomment-930350038)
> 
> They all seem to be related to space charge, which we don't use in the MS.

That is expected, I think, due to solver tolerance issues. If we set them looser, those tests pass with `warpx.use_single_precision_comms=1`

File: Source/Parallelization/WarpXCommUtil.H (hunk 2 above)
Comment (atmyers) [reply Source/Parallelization/WarpXCommUtil.H:7]: Added.

**Comment by RemiLehe** [comment]
@atmyers Could you rebase this PR? Thanks!

File: Examples/Modules/boosted_diags/inputs_3d_slice [hunk 3]
```diff
@@ -122,7 +122,7 @@ btd_openpmd.do_back_transformed_fields = 1
 btd_openpmd.num_snapshots_lab = 4
 btd_openpmd.dz_snapshots_lab = 0.001
 btd_openpmd.fields_to_plot = Ex Ey Ez Bx By Bz jx jy jz rho
-btd_openpmd.format = openpmd
+btd_openpmd.format = plotfile
```
Comment (ax3l) [inline Examples/Modules/boosted_diags/inputs_3d_slice]: Leftover from local testing, maybe:
```suggestion
btd_openpmd.format = openpmd
```

File: Source/Parallelization/WarpXCommUtil.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
+/* Copyright 2019 Andrew Myers
```
Comment (ax3l) [inline Source/Parallelization/WarpXCommUtil.cpp]:
```suggestion
/* Copyright 2021 Andrew Myers
```

File: Source/WarpX.H [hunk 5]
```diff
@@ -163,6 +163,8 @@ public:
     // default is false for standard PSATD and true for Galilean PSATD (set in WarpX.cpp)
     bool update_with_rho = false;
 
+    static int do_single_precision_comms;
```
Comment (ax3l) [inline Source/WarpX.H:167]:
```suggestion
    //! perform field communications in single precision
    static int do_single_precision_comms;
```

File: Docs/source/usage/parameters.rst [hunk 6]
```diff
@@ -192,6 +192,10 @@ Setting up the field mesh
     automatically set so that it is one cell larger than
     ``n_current_deposition_buffer``, on the fine grid.
 
+* ``warpx.do_single_precision_comms`` (`integer`; 0 by default)
+    Perform MPI communications for guard regions in single precision.
```
Comment (ax3l) [inline Docs/source/usage/parameters.rst]:
```suggestion
    Perform MPI communications for field guard regions in single precision.
```

**Review by ax3l — APPROVED** [review]

**Pull request merged**